src/unistd.rs

```diff
@@ -2459,6 +2459,45 @@ impl From<*mut libc::passwd> for User {
 }
 
 impl User {
+    fn from_anything(f: impl Fn(*mut libc::passwd,
+                                *mut libc::c_char,
+                                libc::size_t,
+                                *mut *mut libc::passwd) -> libc::c_int)
+       -> Result<Option<Self>>
+    {
+        let bufsize = match sysconf(SysconfVar::GETPW_R_SIZE_MAX) {
+            Ok(Some(n)) => n as usize,
+            Ok(None) | Err(_) => 1024 as usize,
+        };
+
+        let mut cbuf = Vec::with_capacity(bufsize);
+        let mut pwd = mem::MaybeUninit::<libc::passwd>::uninit();
+        let mut res = ptr::null_mut();
+
+        loop {
+            let error = unsafe {
+                Errno::clear();
+                f(pwd.as_mut_ptr(), cbuf.as_mut_ptr(), cbuf.capacity(), &mut res)
+            };
+
+            if error == 0 {
+                if res.is_null() {
+                    return Ok(None);
+                } else {
+                    let pwd = unsafe { pwd.assume_init() };
+                    return Ok(Some(User::from(&pwd)));
+                }
+            } else if Errno::last() == Errno::ERANGE {
+                // Trigger the internal buffer resizing logic of `Vec` by requiring
+                // more space than the current capacity.
+                unsafe { cbuf.set_len(cbuf.capacity()); }
+                cbuf.reserve(1);
```

Comment:
Extending by only one byte at a time could take awhile.  It would be better to double the storage instead.  That's what the C library does.  Also, you shouldn't extend the buffer ad infinitum.  At some point you should give up and return `ERANGE`.  The C library does that at 1MB.

Reply:
> The link from inside of Vec is very different. In that case, it really is only 1 additional element's worth of space that is needed. Here, an unknown amount of additional space is needed. That's why I suggest doubling the storage. Nix's getcwd() method should probably do the same.

From reading the std code, it seems different. See:

https://github.com/rust-lang/rust/blob/bdfd698f37184da42254a03ed466ab1f90e6fb6c/src/liballoc/raw_vec.rs#L494-L500

and 

https://github.com/rust-lang/rust/blob/bdfd698f37184da42254a03ed466ab1f90e6fb6c/src/liballoc/raw_vec.rs#L421-L434

So it seems todo exactly this; it seems good to keep the code simple but if really desired, I can rework this.

Reply:
That's just the current implementation.  Nix shouldn't rely upon it.  We should actually try to reserve a reasonable amount of space for what our application needs.